---
title_full: "JBrowseR: An R Interface to the JBrowse 2 Genome Browser"
title_short: "JBrowseR: R Interface to JBrowse 2"

# one of `Original Paper`, `Discovery Note`, `Application Note`, `Review`, `Letter to the Editor`, `Conference Paper`    
subject_section: Application Note

access_date:
  - day: DD
    month: MM
    year: YY
      
author:
  - name: Elliot Hershberg
    affiliation: "1"
  - name: Garrett Stevens
    affiliation: "1"
  - name: Colin Diesh
    affiliation: "1"
  - name: Peter Xie
    affiliation: "1"
  - name: Teresa De Jesus Martinez
    affiliation: "1"
  - name: Rob Buels
    affiliation: "1"
  - name: Lincoln Stein
    affiliation: "2"
  - name: Ian Holmes
    affiliation: "1*"
    
address:
  - code: 1
    address: Department of Bioengineering, University of California, Berkeley, Berkeley, CA 94720, USA
  - code: 2
    address: Ontario Institute for Cancer Research, Toronto, ON M5G 0A3, Canada
  
# Full email address to be given, preferably an institution email address.  
corresponding_author_text: "*To whom correspondence should be addressed. E-mail: ihh\\@berkeley.edu"
corresponding_author_email: "ihh\\@berkeley.edu"
  
motivation: |
  Genome browsers are an essential tool in genome analysis. Modern genome browsers enable complex and interactive visualization of a wide variety of genomic data modalities. While such browsers are very powerful, they can be challenging to configure and program for bioinformaticians lacking expertise in web development.
  
results: |
  We have developed an R package that provides an interface to the JBrowse 2 genome browser. The package can be used to configure and customize the browser entirely with R code. The browser can be deployed from the R console, or embedded in Shiny applications or R Markdown documents.

availability: |
  JBrowseR is available for download from CRAN, and the source code is openly available from the Github repository at

supplementary_info: |
  Supplementary data are available at Bioinformatics Online.

output:
  rticles::bioinformatics_article:
    citation_package: natbib
    
bibliography: bibliography.bib
---

Introduction
============

The development of genome browsers is widely considered to be one of the fundamental milestones of the
genomic revolution [@packer2007clickable].
Genome browsers provide researchers the ability to visually display and explore genomic annotations and
data.
Due to their widespread adoption and use, the linear display of genomic information along reference coordinates
is one of the most common represenations of biological data in the 21st century.

Since their original development during the advent of genome sequencing [@kent2002human; @birney2004overview],
genome browsers have made considerable gains in performance and sophistication.
One important development has been the implementation of genome browsers in JavaScript, beginning with
JBrowse [@buels2016jbrowse].
Leveraging JavaScript makes it possible to move computation that previously took place on a server
into the client browser.
Another core advantage of JavaScript based browsers is that they can leverage modern web technologies such
as Canvas and SVG, providing a more responsive and interactive experience for the user.

More recently, JBrowse has been rewritten using newer web technologies such as ReactJS and TypeScript to
create an extensible platform for visualizing and integrating biological data called JBrowse 2.
The platform can be configured and deployed with custom data and settings.
This architecture enables research communities to develop and maintain curated sets of resources and data on 
the web, such as WormBase for the *C. elegans* community [@harris2010wormbase].
A new product offering of JBrowse 2 is a React component that renders a configurable genome browser,
enabling researchers to embed custom browsers into existing React applications.

While the JBrowse 2 React component is powerful and extensible, it can present a steep learning curve
for bioinformaticians who don't have experience with React development.
On the other hand, the R programming language and environment is widely used in the bioinformatics community, 
as evidenced by the size and and usage of efforts such as Bioconductor [@huber2015orchestrating].
To bridge this gap, we introduce JBrowseR, an R interface the JBrowse 2 genome browser. 
JBrowseR is an R package with functions for embedding a custom browser instance in a Shiny app, 
R Markdown document, or the R console.

Materials and Methods
============

Here is how to include math equations in the document (bounded by `$$`):

$$
\begin{aligned}
(x+y)^3&=(x+y)(x+y)^2\\
       &=(x+y)(x^2+2xy+y^2) \label{eqn:example} \\
       &=x^3+3x^2y+3xy^3+x^3. 
\end{aligned}
$$

Describe the approach. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

<!--
problems with longtable and 2-column format. see pnas_template and https://github.com/jgm/pandoc/issues/1023

  Species                    CBS     CV     G3
  ----------------------- ------ ------ ------
  1\. Acetaldehyde           0.0    0.0    0.0
  2\. Vinyl alcohol          9.1    9.6   13.5
  3\. Hydroxyethylidene     50.8   51.2   54.0

  : Comparison of the fitted potential energy surfaces and ab initio
  benchmark electronic energy calculations
-->

Describe the approach. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Describe the approach. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Describe the approach. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

```{r figure, echo = FALSE, fig.width = 3.5, fig.cap = 'Figure from an Rmd chunk.'}
plot(rnorm(10000))
```

Methods
============

Detailed methods. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Detailed methods. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Detailed methods. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Sub-Method
-------------

Details for Method 1. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Details for Method 1. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Details for Method 1. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Details for Method 1. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Method 2
-------------

Details for Method 2. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Details for Method 2. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Details for Method 2. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Details for Method 2. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Discussion
=============

Discussion of results. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Discussion of results. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Discussion of results. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Discussion of results. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Conclusion
=============

Anything else? Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Anything else? Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Anything else? Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Anything else? Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum. Introduce your topic. Lorem ipsum ad nauseum.

Acknowledgements {.unnumbered}
==============

These should be included at the end of the text and not in footnotes. Please ensure you acknowledge all sources of funding, see funding section below.

Details of all funding sources for the work in question should be given in a separate section entitled 'Funding'. This should appear before the 'Acknowledgements' section.

Funding {.unnumbered}
==============
The following rules should be followed:

- The sentence should begin: ‘This work was supported by …’ -
- The full official funding agency name should be given, i.e. ‘National Institutes of Health’, not ‘NIH’ (full RIN-approved list of UK funding agencies)
- Grant numbers should be given in brackets as follows: ‘[grant number xxxx]’
- Multiple grant numbers should be separated by a comma as follows: ‘[grant numbers xxxx, yyyy]’
- Agencies should be separated by a semi-colon (plus ‘and’ before the last funding agency) 
- Where individuals need to be specified for certain sources of funding the following text should be added after the relevant agency or grant number 'to [author initials]'.

An example is given here: ‘This work was supported by the National Institutes of Health [AA123456 to C.S., BB765432 to M.H.]; and the Alcohol & Education Research Council [hfygr667789].’

Oxford Journals will deposit all NIH-funded articles in PubMed Central. See Depositing articles in repositories – information for authors for details. Authors must ensure that manuscripts are clearly indicated as NIH-funded using the guidelines above.

